Liquibase fails when changelog file is not the default resource path

### Describe the bug

When I try to build a native app which uses Liquibase and the changelog file path is not the default the build fails.

### Expected behavior

Native app successfully built.

### Actual behavior

The build process fails with a message:

```
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.11.2:build (default) on project liquibase-quickstart: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.liquibase.deployment.LiquibaseProcessor#nativeImageConfiguration threw an exception: java.lang.IllegalStateException: liquibase.exception.ChangeLogParseException: The file local/changeLog.xml was not found in the configured search path:
```

### How to Reproduce?

[liquibase-quickstart.zip](https://github.com/user-attachments/files/15859145/liquibase-quickstart.zip)

1. Download the liquibase-quickstart.zip.
2. Unzip it where is most convenient for you.
3. Go to the folder created.
4. Open a terminal from there.
5. Run the command, "./mvnw clean install" (it should run with no errors)
6. Build a native image using the command, "./mvnw clean package -Dnative"

### Output of `uname -a` or `ver`

Linux L-PW03AH72 5.15.153.1-microsoft-standard-WSL2 #1 SMP Fri Mar 29 23:14:13 UTC 2024 x86_64 GNU/Linux

### Output of `java -version`

Picked up JAVA_TOOL_OPTIONS: -Dos.version=arch openjdk version "17.0.5" 2022-10-18 LTS OpenJDK Runtime Environment Zulu17.38+21-CA (build 17.0.5+8-LTS) OpenJDK 64-Bit Server VM Zulu17.38+21-CA (build 17.0.5+8-LTS, mixed mode, sharing)

### Mandrel or GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.11.2

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)

### Additional information

I also tried to add additional resource files through the property "quarkus.native.resources.includes". For instance, "quarkus.native.resources.includes=db/changelog/**". But it also did not work.

However, when I created the file "src/main/resources/local/changeLog.xml" and removed the quarkus property, "quarkus.liquibase.search-path=db/changelog", it worked.

Furthermore, if I let the quarkus's property,"quarkus.liquibase.search-path=db/changelog", in the application.properties file it finds twice the same file and fail.

Therefore, I hope to have the same behavior, once set the quarkus.liquibase.search-path it should be handled in the same way no matter the sort of the build.

--- a/extensions/liquibase/deployment/src/main/java/io/quarkus/liquibase/deployment/LiquibaseProcessor.java
+++ b/extensions/liquibase/deployment/src/main/java/io/quarkus/liquibase/deployment/LiquibaseProcessor.java
@@ -3,6 +3,7 @@ package io.quarkus.liquibase.deployment;
 import static io.quarkus.deployment.annotations.ExecutionTime.STATIC_INIT;
 import static java.util.function.Predicate.not;
 
+import java.io.FileNotFoundException;
 import java.io.IOException;
 import java.nio.file.Paths;
 import java.util.ArrayList;
@@ -61,8 +62,10 @@ import io.quarkus.deployment.util.ServiceUtil;
 import io.quarkus.liquibase.LiquibaseDataSource;
 import io.quarkus.liquibase.LiquibaseFactory;
 import io.quarkus.liquibase.runtime.LiquibaseBuildTimeConfig;
+import io.quarkus.liquibase.runtime.LiquibaseDataSourceBuildTimeConfig;
 import io.quarkus.liquibase.runtime.LiquibaseFactoryProducer;
 import io.quarkus.liquibase.runtime.LiquibaseRecorder;
+import io.quarkus.runtime.util.StringUtil;
 import liquibase.change.Change;
 import liquibase.change.DatabaseChangeProperty;
 import liquibase.change.core.CreateProcedureChange;
@@ -77,6 +80,9 @@ import liquibase.parser.ChangeLogParser;
 import liquibase.parser.ChangeLogParserFactory;
 import liquibase.plugin.AbstractPluginFactory;
 import liquibase.resource.ClassLoaderResourceAccessor;
+import liquibase.resource.CompositeResourceAccessor;
+import liquibase.resource.DirectoryResourceAccessor;
+import liquibase.resource.ResourceAccessor;
 
 class LiquibaseProcessor {
 
@@ -360,57 +366,89 @@ class LiquibaseProcessor {
             return Collections.emptyList();
         }
 
-        ChangeLogParameters changeLogParameters = new ChangeLogParameters();
+        List<LiquibaseDataSourceBuildTimeConfig> liquibaseDataSources = new ArrayList<>();
 
-        ChangeLogParserFactory changeLogParserFactory = ChangeLogParserFactory.getInstance();
+        if (DataSourceUtil.hasDefault(dataSourceNames)) {
+            liquibaseDataSources.add(liquibaseBuildConfig.defaultDataSource);
+        }
 
-        Set<String> resources = new LinkedHashSet<>();
-
-        ClassLoaderResourceAccessor classLoaderResourceAccessor = new ClassLoaderResourceAccessor(
-                Thread.currentThread().getContextClassLoader());
-
-        try {
-            // default datasource
-            if (DataSourceUtil.hasDefault(dataSourceNames)) {
-                resources.addAll(findAllChangeLogFiles(liquibaseBuildConfig.defaultDataSource.changeLog, changeLogParserFactory,
-                        classLoaderResourceAccessor, changeLogParameters));
-            }
-
-            // named datasources
-            Collection<String> namedDataSourceChangeLogs = dataSourceNames.stream()
-                    .filter(n -> !DataSourceUtil.isDefault(n))
-                    .map(liquibaseBuildConfig::getConfigForDataSourceName)
-                    .map(c -> c.changeLog)
-                    .collect(Collectors.toCollection(LinkedHashSet::new));
-
-            for (String namedDataSourceChangeLog : namedDataSourceChangeLogs) {
-                resources.addAll(
-                        findAllChangeLogFiles(namedDataSourceChangeLog, changeLogParserFactory, classLoaderResourceAccessor,
-                                changeLogParameters));
-            }
-
-            LOGGER.debugf("Liquibase changeLogs: %s", resources);
-
-            return new ArrayList<>(resources);
-
-        } finally {
-            try {
-                classLoaderResourceAccessor.close();
-            } catch (Exception ignored) {
-                // close() really shouldn't declare that exception, see also https://github.com/liquibase/liquibase/pull/2576
+        for (String dataSourceName : dataSourceNames) {
+            if (!DataSourceUtil.isDefault(dataSourceName)) {
+                liquibaseDataSources.add(liquibaseBuildConfig.getConfigForDataSourceName(dataSourceName));
             }
         }
+
+        ChangeLogParameters changeLogParameters = new ChangeLogParameters();
+        ChangeLogParserFactory changeLogParserFactory = ChangeLogParserFactory.getInstance();
+        Set<String> resources = new LinkedHashSet<>();
+        for (LiquibaseDataSourceBuildTimeConfig liquibaseDataSourceConfig : liquibaseDataSources) {
+
+            Optional<List<String>> oSearchPaths = liquibaseDataSourceConfig.searchPath;
+            String changeLog = liquibaseDataSourceConfig.changeLog;
+            String parsedChangeLog = parseChangeLog(oSearchPaths, changeLog);
+
+            try (ResourceAccessor resourceAccessor = resolveResourceAccessor(oSearchPaths, changeLog)) {
+                resources.addAll(findAllChangeLogFiles(parsedChangeLog, changeLogParserFactory,
+                        resourceAccessor, changeLogParameters));
+            } catch (Exception ex) {
+                throw new IllegalStateException(ex);
+            }
+        }
+
+        LOGGER.debugf("Liquibase changeLogs: %s", resources);
+        return new ArrayList<>(resources);
+    }
+
+    private ResourceAccessor resolveResourceAccessor(Optional<List<String>> oSearchPaths, String changeLog)
+            throws FileNotFoundException {
+
+        CompositeResourceAccessor compositeResourceAccessor = new CompositeResourceAccessor();
+        compositeResourceAccessor
+                .addResourceAccessor(new ClassLoaderResourceAccessor(Thread.currentThread().getContextClassLoader()));
+
+        if (!changeLog.startsWith("filesystem:") && oSearchPaths.isEmpty()) {
+            return compositeResourceAccessor;
+        }
+
+        if (oSearchPaths.isEmpty()) {
+            compositeResourceAccessor.addResourceAccessor(
+                    new DirectoryResourceAccessor(
+                            Paths.get(StringUtil.changePrefix(changeLog, "filesystem:", "")).getParent()));
+            return compositeResourceAccessor;
+        }
+
+        for (String searchPath : oSearchPaths.get()) {
+            compositeResourceAccessor.addResourceAccessor(new DirectoryResourceAccessor(Paths.get(searchPath)));
+        }
+
+        return compositeResourceAccessor;
+    }
+
+    private String parseChangeLog(Optional<List<String>> oSearchPaths, String changeLog) {
+
+        if (changeLog.startsWith("filesystem:") && oSearchPaths.isEmpty()) {
+            return Paths.get(StringUtil.changePrefix(changeLog, "filesystem:", "")).getFileName().toString();
+        }
+
+        if (changeLog.startsWith("filesystem:")) {
+            return StringUtil.changePrefix(changeLog, "filesystem:", "");
+        }
+
+        if (changeLog.startsWith("classpath:")) {
+            return StringUtil.changePrefix(changeLog, "classpath:", "");
+        }
+
+        return changeLog;
     }
 
     /**
      * Finds all resource files for the given change log file
      */
     private Set<String> findAllChangeLogFiles(String file, ChangeLogParserFactory changeLogParserFactory,
-            ClassLoaderResourceAccessor classLoaderResourceAccessor,
-            ChangeLogParameters changeLogParameters) {
+            ResourceAccessor resourceAccessor, ChangeLogParameters changeLogParameters) {
         try {
-            ChangeLogParser parser = changeLogParserFactory.getParser(file, classLoaderResourceAccessor);
-            DatabaseChangeLog changelog = parser.parse(file, changeLogParameters, classLoaderResourceAccessor);
+            ChangeLogParser parser = changeLogParserFactory.getParser(file, resourceAccessor);
+            DatabaseChangeLog changelog = parser.parse(file, changeLogParameters, resourceAccessor);
 
             if (changelog != null) {
                 Set<String> result = new LinkedHashSet<>();

